README
NAME
    GnuPG::Interface - Perl interface to GnuPG

SYNOPSIS
      # A simple example
      use IO::Handle;
      use GnuPG::Interface;

      # setting up the situation
      my $gnupg = GnuPG::Interface->new();
      $gnupg->options->hash_init( armor   => 1,
                                  homedir => '/home/foobar' );

      # Note you can set the recipients even if you aren't encrypting!
      $gnupg->options->push_recipients( 'ftobin@cpan.org' );
      $gnupg->options->meta_interactive( 0 );

      # how we create some handles to interact with GnuPG
      my $input   = IO::Handle->new();
      my $output  = IO::Handle->new();
      my $handles = GnuPG::Handles->new( stdin  => $input,
                                         stdout => $output );

      # Now we'll go about encrypting with the options already set
      my @plaintext = ( 'foobar' );
      my $pid = $gnupg->encrypt( handles => $handles );

      # Now we write to the input of GnuPG
      print $input @plaintext;
      close $input;

      # now we read the output
      my @ciphertext = <$output>;
      close $output;

      waitpid $pid, 0;

DESCRIPTION
    GnuPG::Interface and its associated modules are designed to provide an
    object-oriented method for interacting with GnuPG, being able to perform
    functions such as but not limited to encrypting, signing, decryption,
    verification, and key-listing parsing.

  How Data Member Accessor Methods are Created
    Each module in the GnuPG::Interface bundle relies on Moo to generate the
    get/set methods used to set the object's data members. *This is very
    important to realize.* This means that any data member which is a list
    has special methods assigned to it for pushing, popping, and clearing
    the list.

  Understanding Bidirectional Communication
    It is also imperative to realize that this package uses interprocess
    communication methods similar to those used in IPC::Open3 and
    "Bidirectional Communication with Another Process" in perlipc, and that
    users of this package need to understand how to use this method because
    this package does not abstract these methods for the user greatly. This
    package is not designed to abstract this away entirely (partly for
    security purposes), but rather to simply help create 'proper', clean
    calls to GnuPG, and to implement key-listing parsing. Please see
    "Bidirectional Communication with Another Process" in perlipc to learn
    how to deal with these methods.

    Using this package to do message processing generally invovlves creating
    a GnuPG::Interface object, creating a GnuPG::Handles object, setting
    some options in its options data member, and then calling a method which
    invokes GnuPG, such as clearsign. One then interacts with with the
    handles appropriately, as described in "Bidirectional Communication with
    Another Process" in perlipc.

GnuPG Versions
    As of this version of GnuPG::Interface, there are two supported versions
    of GnuPG: 1.4.x and 2.2.x. The GnuPG download page
    <https://gnupg.org/download/index.html> has updated information on the
    currently supported versions.

    GnuPG released 2.0 and 2.1 versions in the past and some packaging
    systems may still provide these if you install the default "gpg",
    "gnupg", "gnupg2", etc. packages. This modules supports only version
    2.2.x, so you may need to find additional package repositories or build
    from source to get the updated version.

OBJECT METHODS
  Initialization Methods
    new( *%initialization_args* )
        This methods creates a new object. The optional arguments are
        initialization of data members.

    hash_init( *%args* ).

  Object Methods which use a GnuPG::Handles Object
    list_public_keys( % )
    list_sigs( % )
    list_secret_keys( % )
    encrypt( % )
    encrypt_symmetrically( % )
    sign( % )
    clearsign( % )
    detach_sign( % )
    sign_and_encrypt( % )
    decrypt( % )
    verify( % )
    import_keys( % )
    export_keys( % )
    recv_keys( % )
    send_keys( % )
    search_keys( % )
        These methods each correspond directly to or are very similar to a
        GnuPG command described in gpg. Each of these methods takes a hash,
        which currently must contain a key of handles which has the value of
        a GnuPG::Handles object. Another optional key is command_args which
        should have the value of an array reference; these arguments will be
        passed to GnuPG as command arguments. These command arguments are
        used for such things as determining the keys to list in the
        export_keys method. *Please note that GnuPG command arguments are
        not the same as GnuPG options*. To understand what are options and
        what are command arguments please read "COMMANDS" in gpg and
        "OPTIONS" in gpg.

        Each of these calls returns the PID for the resulting GnuPG process.
        One can use this PID in a "waitpid" call instead of a "wait" call if
        more precise process reaping is needed.

        These methods will attach the handles specified in the handles
        object to the running GnuPG object, so that bidirectional
        communication can be established. That is, the optionally-defined
        stdin, stdout, stderr, status, logger, and passphrase handles will
        be attached to GnuPG's input, output, standard error, the handle
        created by setting status-fd, the handle created by setting
        logger-fd, and the handle created by setting passphrase-fd
        respectively. This tying of handles of similar to the process done
        in *IPC::Open3*.

        If you want the GnuPG process to read or write directly to an
        already-opened filehandle, you cannot do this via the normal
        *IPC::Open3* mechanisms. In order to accomplish this, set the
        appropriate handles data member to the already-opened filehandle,
        and then set the option direct to be true for that handle, as
        described in "options" in GnuPG::Handles. For example, to have GnuPG
        read from the file input.txt and write to output.txt, the following
        snippet may do:

          my $infile  = IO::File->new( 'input.txt' );
          my $outfile = IO::File->new( '>output.txt' );
          my $handles = GnuPG::Handles->new( stdin  => $infile,
                                             stdout => $outfile,
                                           );
          $handles->options( 'stdin'  )->{direct} = 1;
          $handles->options( 'stdout' )->{direct} = 1;

        If any handle in the handles object is not defined, GnuPG's input,
        output, and standard error will be tied to the running program's
        standard error, standard output, or standard error. If the status or
        logger handle is not defined, this channel of communication is never
        established with GnuPG, and so this information is not generated and
        does not come into play.

        If the passphrase data member handle of the handles object is not
        defined, but the the passphrase data member handle of
        GnuPG::Interface object is, GnuPG::Interface will handle passing
        this information into GnuPG for the user as a convenience. Note that
        this will result in GnuPG::Interface storing the passphrase in
        memory, instead of having it simply 'pass-through' to GnuPG via a
        handle.

        If neither the passphrase data member of the GnuPG::Interface nor
        the passphrase data member of the handles object is defined, then
        GnuPG::Interface assumes that access and control over the secret key
        will be handled by the running gpg-agent process. This represents
        the simplest mode of operation with the GnuPG "stable" suite
        (version 2.2 and later). It is also the preferred mode for tools
        intended to be user-facing, since the user will be prompted directly
        by gpg-agent for use of the secret key material. Note that for
        programmatic use, this mode requires the gpg-agent and pinentry to
        already be correctly configured.

  Other Methods
    get_public_keys( @search_strings )
    get_secret_keys( @search_strings )
    get_public_keys_with_sigs( @search_strings )
        These methods create and return objects of the type GnuPG::PublicKey
        or GnuPG::SecretKey respectively. This is done by parsing the output
        of GnuPG with the option with-colons enabled. The objects created do
        or do not have signature information stored in them, depending if
        the method ends in *_sigs*; this separation of functionality is
        there because of performance hits when listing information with
        signatures.

    test_default_key_passphrase()
        This method will return a true or false value, depending on whether
        GnuPG reports a good passphrase was entered while signing a short
        message using the values of the passphrase data member, and the
        default key specified in the options data member.

    version()
        Returns the version of GnuPG that GnuPG::Interface is running.

Invoking GnuPG with a custom call
    GnuPG::Interface attempts to cover a lot of the commands of GnuPG that
    one would want to perform; however, there may be a lot more calls that
    GnuPG is and will be capable of, so a generic command interface is
    provided, "wrap_call".

    wrap_call( %args )
        Call GnuPG with a custom command. The %args hash must contain at
        least the following keys:

        commands
            The value of this key in the hash must be a reference to a a
            list of commands for GnuPG, such as "[ qw( --encrypt --sign )
            ]".

        handles
            As with most other GnuPG::Interface methods, handles must be a
            GnuPG::Handles object.

        The following keys are optional.

        command_args
            As with other GnuPG::Interface methods, the value in hash for
            this key must be a reference to a list of arguments to be passed
            to the GnuPG command, such as which keys to list in a
            key-listing.

OBJECT DATA MEMBERS
    call
        This defines the call made to invoke GnuPG. Defaults to 'gpg'; this
        should be changed if 'gpg' is not in your path, or there is a
        different name for the binary on your system.

    passphrase
        In order to lessen the burden of using handles by the user of this
        package, setting this option to one's passphrase for a secret key
        will allow the package to enter the passphrase via a handle to GnuPG
        by itself instead of leaving this to the user. See also "passphrase"
        in GnuPG::Handles.

    options
        This data member, of the type GnuPG::Options; the setting stored in
        this data member are used to determine the options used when calling
        GnuPG via *any* of the object methods described in this package. See
        GnuPG::Options for more information.

EXAMPLES
    The following setup can be done before any of the following examples:

      use IO::Handle;
      use GnuPG::Interface;

      my @original_plaintext = ( "How do you doo?" );
      my $passphrase = "Three Little Pigs";

      my $gnupg = GnuPG::Interface->new();

      $gnupg->options->hash_init( armor    => 1,
                                  recipients => [ 'ftobin@uiuc.edu',
                                                  '0xABCD1234ABCD1234ABCD1234ABCD1234ABCD1234' ],
                                  meta_interactive => 0 ,
                                );

       $gnupg->options->debug_level(4);

       $gnupg->options->logger_file("/tmp/gnupg-$$-decrypt-".time().".log");

  Encrypting
      # We'll let the standard error of GnuPG pass through
      # to our own standard error, by not creating
      # a stderr-part of the $handles object.
      my ( $input, $output ) = ( IO::Handle->new(),
                                 IO::Handle->new() );

      my $handles = GnuPG::Handles->new( stdin    => $input,
                                         stdout   => $output );

      # this sets up the communication
      # Note that the recipients were specified earlier
      # in the 'options' data member of the $gnupg object.
      my $pid = $gnupg->encrypt( handles => $handles );

      # this passes in the plaintext
      print $input @original_plaintext;

      # this closes the communication channel,
      # indicating we are done
      close $input;

      my @ciphertext = <$output>;  # reading the output

      waitpid $pid, 0;  # clean up the finished GnuPG process

  Signing
      # This time we'll catch the standard error for our perusing
      my ( $input, $output, $error ) = ( IO::Handle->new(),
                                         IO::Handle->new(),
                                         IO::Handle->new(),
                                       );

      my $handles = GnuPG::Handles->new( stdin    => $input,
                                         stdout   => $output,
                                         stderr   => $error,
                                       );

      # indicate our pasphrase through the
      # convenience method
      $gnupg->passphrase( $passphrase );

      # this sets up the communication
      my $pid = $gnupg->sign( handles => $handles );

      # this passes in the plaintext
      print $input @original_plaintext;

      # this closes the communication channel,
      # indicating we are done
      close $input;

      my @ciphertext   = <$output>;  # reading the output
      my @error_output = <$error>;   # reading the error

      close $output;
      close $error;

      waitpid $pid, 0;  # clean up the finished GnuPG process

  Decryption
      # This time we'll catch the standard error for our perusing
      # as well as passing in the passphrase manually
      # as well as the status information given by GnuPG
      my ( $input, $output, $error, $passphrase_fh, $status_fh )
        = ( IO::Handle->new(),
            IO::Handle->new(),
            IO::Handle->new(),
            IO::Handle->new(),
            IO::Handle->new(),
          );

      my $handles = GnuPG::Handles->new( stdin      => $input,
                                         stdout     => $output,
                                         stderr     => $error,
                                         passphrase => $passphrase_fh,
                                         status     => $status_fh,
                                       );

      # this time we'll also demonstrate decrypting
      # a file written to disk
      # Make sure you "use IO::File" if you use this module!
      my $cipher_file = IO::File->new( 'encrypted.gpg' );

      # this sets up the communication
      my $pid = $gnupg->decrypt( handles => $handles );

      # This passes in the passphrase
      print $passphrase_fh $passphrase;
      close $passphrase_fh;

      # this passes in the plaintext
      print $input $_ while <$cipher_file>;

      # this closes the communication channel,
      # indicating we are done
      close $input;
      close $cipher_file;

      my @plaintext    = <$output>;    # reading the output
      my @error_output = <$error>;     # reading the error
      my @status_info  = <$status_fh>; # read the status info

      # clean up...
      close $output;
      close $error;
      close $status_fh;

      waitpid $pid, 0;  # clean up the finished GnuPG process

  Printing Keys
      # This time we'll just let GnuPG print to our own output
      # and read from our input, because no input is needed!
      my $handles = GnuPG::Handles->new();

      my @ids = ( 'ftobin', '0xABCD1234ABCD1234ABCD1234ABCD1234ABCD1234' );

      # this time we need to specify something for
      # command_args because --list-public-keys takes
      # search ids as arguments
      my $pid = $gnupg->list_public_keys( handles      => $handles,
                                          command_args => [ @ids ] );

       waitpid $pid, 0;

  Creating GnuPG::PublicKey Objects
      my @ids = [ 'ftobin', '0xABCD1234ABCD1234ABCD1234ABCD1234ABCD1234' ];

      my @keys = $gnupg->get_public_keys( @ids );

      # no wait is required this time; it's handled internally
      # since the entire call is encapsulated

  Custom GnuPG call
      # assuming $handles is a GnuPG::Handles object
      my $pid = $gnupg->wrap_call
        ( commands     => [ qw( --list-packets ) ],
          command_args => [ qw( test/key.1.asc ) ],
          handles      => $handles,
        );

        my @out = <$handles->stdout()>;
        waitpid $pid, 0;

FAQ
    How do I get GnuPG::Interface to read/write directly from a filehandle?
        You need to set GnuPG::Handles direct option to be true for the
        filehandles in concern. See "options" in GnuPG::Handles and "Object
        Methods which use a GnuPG::Handles Object" for more information.

    Why do you make it so difficult to get GnuPG to write/read from a
    filehandle? In the shell, I can just call GnuPG with the --outfile
    option!
        There are lots of issues when trying to tell GnuPG to read/write
        directly from a file, such as if the file isn't there, or there is a
        file, and you want to write over it! What do you want to happen
        then? Having the user of this module handle these questions
        beforehand by opening up filehandles to GnuPG lets the user know
        fully what is going to happen in these circumstances, and makes the
        module less error-prone.

    When having GnuPG process a large message, sometimes it just hanges
    there.
        Your problem may be due to buffering issues; when GnuPG reads/writes
        to non-direct filehandles (those that are sent to filehandles which
        you read to from into memory, not that those access the disk),
        buffering issues can mess things up. I recommend looking into
        "options" in GnuPG::Handles.

NOTES
    This package is the successor to PGP::GPG::MessageProcessor, which I
    found to be too inextensible to carry on further. A total redesign was
    needed, and this is the resulting work.

    After any call to a GnuPG-command method of GnuPG::Interface in which
    one passes in the handles, one should all wait to clean up GnuPG from
    the process table.

BUGS
    Currently there are problems when transmitting large quantities of
    information over handles; I'm guessing this is due to buffering issues.
    This bug does not seem specific to this package; IPC::Open3 also appears
    affected.

    I don't know yet how well this modules handles parsing OpenPGP v3 keys.

SEE ALSO
    GnuPG::Options, GnuPG::Handles, GnuPG::PublicKey, GnuPG::SecretKey, gpg,
    "Bidirectional Communication with Another Process" in perlipc

LICENSE
    This module is free software; you can redistribute it and/or modify it
    under the same terms as Perl itself.

AUTHOR
    GnuPG::Interface is currently maintained by Best Practical Solutions
    <BPS@cpan.org>.

    Frank J. Tobin, ftobin@cpan.org was the original author of the package.



Changes
Revision history for GnuPG-Interface

1.03 - 2023-09-14
 - Add fix for running in taint mode for Perl 5.38.0

1.02 - 2021-04-09
 - Work around an issue with CLI options for deleting secret keys in gpg 2.2

1.01 - 2021-01-26
 - Make $gnupg->call('foo') update the internal GnuPG version number.
 - Localize tainted PATH env to make exec happy

1.0 - 2020-05-13
 - Limit support to GnuPG 2.2+ and 1.4
 - Additional information from keys when using GnuPG 2.2 or higher
 - Add support for use of agent/pinentry
 - Updated options to add ignore_mdc_error and logging
 - Improvements to tests
 - Update pubkey_data documentation
 - Special thanks to dkg on Github for a large PR with updates for GnuPG 2
 - Thanks also to ntyni on Github for a pointer to test updates also dealing
   with version changes

0.52 - 2016-02-16
 - Skip "grp" records, generated by GPG 2.1; this suppresses "unknown
   record type" warnings
 - Add explicit Fatal dependency; though nominally part of code perl,
   RedHat's perl does not ship with it
 - Ensure that the trustdb is created before attempting to encrypt; gpg2
   requires that it exist, even for commands with --trust-model=always.
   See https://bugs.debian.org/cgi-bin/bugreport.cgi?bug=751266

0.51 - 2014-12-17
 - Update README file
 - Work around gpg2 bug by omitting --homedir during symmetric
   encryption

0.50 - 2014-03-14
 - Version 0.49 implicitly required Moose; switch to a technique that
   does not
 - Modernize CHANGES

0.49 - 2014-03-13
 - Restore context-sensitive (array/arrayref) behavior of multiple array
   methods from 0.46.
 - Fix MANIFEST/.gitignore inconsistency

0.48 - 2014-03-10
 - Switch from --always-trust to --trust-model=always

0.47 - 2014-03-10
 - No changes from 0.47_02

0.47_02 - 2014-02-14
 - Remove a stray 'use Data::Dumper::Concise' added in 0.47_01

0.47_01 - 2014-01-27
 - Switch from Any::Moose to Moo
 - Accept "gpg (GnuPG/MacGPG2)" as a valid gpg version
 - Typo fixes in documentation

0.46 - 2012-10-25
 - Add a ->search_keys method
 - Add a ->version method
 - Remove dead code for finding gnupg2 binary

0.45 - 2011-10-26
 - Include trailing columns when parsing --fixed-list-mode output

0.44 - 2011-05-02
 - Bump Math::BigInt dependency to get the new 'try GMP' syntax.

0.43 - 2011-03-08
 - Stable release

0.42_02 - 2010-06-05
 - Additional cleanups from dkg

0.42_01 - 2010-05-10
 - GnuPG::Revoker: improve docs, compare() should fail if the signature
   counts differ - dkg
 - Handle revoker packets (rvk) - dkg
 - Move compare() into GnuPG::Signature, got rid of
   t/GnuPG/ComparableSignature.pm - dkg
 - Move signature comparison into ComparableKey.pm instead of
   ComparableSubKey.pm - dkg
 - Move fingerprint comparison directly into GnuPG::Fingerprint - dkg
 - Change around some variable names for consistency's sake:
      $current_key becomes $current_primary_key
      $current_fingerprinted_key becomes $current_key  -dkg
 - Fixed synopsis example in GnuPG::Signature pod - dkg
 - Allow for primary key to have per-key (useful for signatures of class
   0x1f, see http://tools.ietf.org/html/rfc4880#section-5.2.1) - dkg
 - Add revocations to keys and user ids and user attributes - dkg
 - Add signature class and exportability to GnuPG::Signature - dkg
 - Introduced GnuPG::UserAttribute to handle uat packets - dkg
 - Actually check validity of signatures and report them - dkg
 - Support more than 1 signature over each subkey - dkg
 - Do not bother shipping test/random_seed - dkg
 - Fix copy method of GnuPG::Options:
     The result of not checking for definedness here is to never copy the
     meta_immutable value successfully (as that is true by default).  This led
     to a FTBFS (failure to build from source) when running non-interactively.
     See also:  http://bugs.debian.org/549743  - Tim Retout
 - Expose signature expiration times - dkg
 - Take advantage of --fixed-list-mode and report timestamps at 1Hz
   precision instead of daily precision - dkg
 - Always use --fixed-list-mode for consistency and better granularity
   of data - dkg
 - Unescape strings to handle User IDs with colons in them - dkg
 - Add usage_flags to keys - dkg
 - Several doc patches from Daniel Kahn Gillmor
 - Fix for documented typos reported by SYSMON Fixes rt.cpan.org#50377
 - Fix POD errors - alexmv

0.42 - 2009-09-30
 - Support for GPG2

0.41_01 - 2009-09-25
 - Beginnings of support for GPG2

0.40_04 - 2009-04-21
 - Use Any::Moose instead of Moose for Mouse celerity (Sartak)

0.40_1 - 2008-11-15
 - [rt.cpan.org #40963] Replace Class::MethodMaker with Moose (Chris
   Prather)

0.36 - 2007-08-13
 - [rt.cpan.org #28814] - Performance improvement from mehradek
   (Radoslaw Zielinski)
        -use English;
        +use English qw( -no_match_vars );

0.35 - 2007-04-20
 - New Maintainer: Jesse Vincent <jesse@bestpractical.com> took over
   maintenance of this module.
 - Removed test key expiry dates. (Fixes
   http://rt.cpan.org/Ticket/Display.html?id=17618)
 - Applied secret key output patch for modern GPG from
   http://rt.cpan.org/Ticket/Display.html?id=17619
 - Applied patch to support 'tru' record types from
   (http://search.cpan.org/src/JRED/Mail-GPG-1.0.6/patches/)

0.34 - 2002-09-26
 - Documentation fixes.

0.33 - 2002-06-14
 - Fixed a bug in GnuPG::Interface->import_keys() so that it doesn't
   overwrite your 'command_args' anymore.  Thanks to Peter Palfrader
   <weasel@debian.org> for pointing this out.

0.32 - 2002-06-11
 - Extended the expiration on the test keys another 4 years so that the
   test suite works.
 - Documentation fixes.
 - Other small cleanups.

0.31 - 2001-05-03
 - Fixed stalling test cases.
 - Added deprecation support for fields of GnuPG::Interface::wrap_call

0.30 - 2001-05-01
 - Re-worked inheritance tree so that GnuPG::SecretKey and
   GnuPG::PublicKey are sub-classes of newly-added GnuPG::PrimaryKey.
 - Tested with GnuPG 1.0.5.
 - GnuPG::Fingerprint deprecate hex_data(), in favor of as_hex_string().

lib/GnuPG/Options.pm
#  Options.pm
#    - providing an object-oriented approach to GnuPG's options
#
#  Copyright (C) 2000 Frank J. Tobin <ftobin@cpan.org>
#
#  This module is free software; you can redistribute it and/or modify it
#  under the same terms as Perl itself.
#
#  This program is distributed in the hope that it will be useful,
#  but WITHOUT ANY WARRANTY; without even the implied warranty of
#  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
#
#  $Id: Options.pm,v 1.14 2001/08/21 13:31:50 ftobin Exp $
#

package GnuPG::Options;
use Moo;
use MooX::late;
use MooX::HandlesVia;
with qw(GnuPG::HashInit);

use constant BOOLEANS => qw(
    armor
    no_greeting
    verbose
    no_verbose
    quiet
    batch
    always_trust
    rfc1991
    openpgp
    force_v3_sigs
    no_options
    textmode
    meta_pgp_5_compatible
    meta_pgp_2_compatible
    meta_interactive
    ignore_mdc_error
);

use constant SCALARS => qw(
    homedir
    default_key
    comment
    status_fd
    logger_fd
    passphrase_fd
    command_fd
    compress_algo
    options
    meta_signing_key
    meta_signing_key_id
    debug_level
    logger_file
);

use constant LISTS => qw(
    encrypt_to
    recipients
    meta_recipients_keys
    meta_recipients_key_ids
    extra_args
);

has $_ => (
    isa     => 'Bool',
    is      => 'rw',
    clearer => 'clear_' . $_,
) for BOOLEANS;

has $_ => (
    isa     => 'Any',
    is      => 'rw',
    clearer => 'clear_' . $_,
) for SCALARS;

for my $list (LISTS) {
    my $ref = $list . "_ref";
    has $ref => (
        handles_via => 'Array',
        is          => 'rw',
        lazy        => 1,
        clearer     => "clear_$list",
        default     => sub { [] },
        handles     => {
            "push_$list" => 'push',
        },
    );

    no strict 'refs';
    *{$list} = sub {
        my $self = shift;
        return wantarray ? @{$self->$ref(@_)} : $self->$ref(@_);
    };
}

sub BUILD {
    my ( $self, $args ) = @_;
    # Newer GnuPG will force failure for old ciphertext unless set
    $args->{ignore_mdc_error} //= 1;

    $self->hash_init( meta_interactive => 1 );
    $self->hash_init(%$args);
}

sub copy {
    my ($self) = @_;

    my $new = ( ref $self )->new();

    foreach my $field ( BOOLEANS, SCALARS, LISTS ) {
        my $value = $self->$field();
        next unless defined $value;
        $new->$field($value);
    }

    return $new;
}

sub get_args {
    my ($self) = @_;

    return (
        $self->get_meta_args(),
        $self->get_option_args(),
        $self->extra_args(),
    );
}

sub get_option_args {
    my ($self) = @_;

    my @args = ();

    push @args, '--homedir', $self->homedir() if $self->homedir();
    push @args, '--options', $self->options() if $self->options();
    push @args, '--no-options' if $self->no_options();
    push @args, '--armor'      if $self->armor();
    push @args, '--textmode'   if $self->textmode();
    push @args, '--default-key', $self->default_key() if $self->default_key();
    push @args, '--no-greeting'  if $self->no_greeting();
    push @args, '--verbose'      if $self->verbose();
    push @args, '--no-verbose'   if $self->no_verbose();
    push @args, '--quiet'        if $self->quiet();
    push @args, '--batch'        if $self->batch();
    push @args, '--trust-model=always' if $self->always_trust();
    push @args, '--comment', $self->comment() if defined $self->comment();
    push @args, '--force-v3-sigs' if $self->force_v3_sigs();
    push @args, '--rfc1991'       if $self->rfc1991;
    push @args, '--openpgp'       if $self->openpgp();
    push @args, '--compress-algo', $self->compress_algo()
        if defined $self->compress_algo();

    push @args, '--status-fd', $self->status_fd()
        if defined $self->status_fd();
    push @args, '--logger-fd', $self->logger_fd()
        if defined $self->logger_fd();
    push @args, '--passphrase-fd', $self->passphrase_fd()
        if defined $self->passphrase_fd();
    push @args, '--command-fd', $self->command_fd()
        if defined $self->command_fd();

    push @args, map { ( '--recipient',  $_ ) } $self->recipients();
    push @args, map { ( '--encrypt-to', $_ ) } $self->encrypt_to();

    push @args, '--debug-level', $self->debug_level() if ($self->debug_level);
    push @args, '--logger-file', $self->logger_file() if ($self->logger_file());

    push @args, '--ignore-mdc-error' if ($self->ignore_mdc_error());

    return @args;
}

sub get_meta_args {
    my ($self) = @_;

    my @args = ();

    push @args, '--compress-algo', 1, '--force-v3-sigs'
        if $self->meta_pgp_5_compatible();
    push @args, '--rfc1991' if $self->meta_pgp_2_compatible();
    push @args, '--batch', '--no-tty' if not $self->meta_interactive();

    # To eliminate confusion, we'll move to having any options
    # that deal with keys end in _id(s) if they only take
    # an id; otherwise we assume that a GnuPG::Key
    push @args, '--default-key', $self->meta_signing_key_id()
        if $self->meta_signing_key_id();
    push @args, '--default-key', $self->meta_signing_key()->hex_id()
        if $self->meta_signing_key();

    push @args,
        map { ( '--recipient', $_ ) } $self->meta_recipients_key_ids();
    push @args,
        map { ( '--recipient', $_->hex_id() ) } $self->meta_recipients_keys();

    return @args;
}

1;

__END__

=head1 NAME

GnuPG::Options - GnuPG options embodiment

=head1 SYNOPSIS

  # assuming $gnupg is a GnuPG::Interface object
  $gnupg->options->armor( 1 );
  $gnupg->options->push_recipients( 'ftobin', '0xABCD1234ABCD1234ABCD1234ABCD1234ABCD1234' );

=head1 DESCRIPTION

GnuPG::Options objects are generally not instantiated on their
own, but rather as part of a GnuPG::Interface object.

=head1 OBJECT METHODS

=over 4

=item new( I<%initialization_args> )

This methods creates a new object.  The optional arguments are
initialization of data members.

=item hash_init( I<%args> ).


=item copy

Returns a copy of this object.  Useful for 'saving' options.

=item get_args

Returns a list of arguments to be passed to GnuPG based
on data members which are 'meta_' options, regular options,
and then I<extra_args>, in that order.

=back

=head1 OBJECT DATA MEMBERS

=over 4

=item homedir

=item armor

=item textmode

=item default_key

=item no_greeting

=item verbose

=item no_verbose

=item quiet

=item batch

=item always_trust

=item comment

=item status_fd

=item logger_fd

=item passphrase_fd

=item compress_algo

=item force_v3_sigs

=item rfc1991

=item openpgp

=item options

=item no_options

=item encrypt_to

=item recipients

=back

These options correlate directly to many GnuPG options.  For those that
are boolean to GnuPG, simply that argument is passed.  For those
that are associated with a scalar, that scalar is passed passed
as an argument appropriate.  For those that can be specified more
than once, such as B<recipients>, those are considered lists
and passed accordingly.  Each are undefined or false to begin.

=head2 Meta Options

Meta options are those which do not correlate directly to any
option in GnuPG, but rather are generally a bundle of options
used to accomplish a specific goal, such as obtaining
compatibility with PGP 5.  The actual arguments each of these
reflects may change with time.  Each defaults to false unless
otherwise specified.

These options are being designed and to provide a non-GnuPG-specific
abstraction, to help create compatibility with a possible
PGP::Interface module.

To help avoid confusion, methods with take a form of a key as
an object shall be prepended with I<_id(s)> if they only
take an id; otherwise assume an object of type GnuPG::Key
is required.

=over 4

=item meta_pgp_5_compatible

If true, arguments are generated to try to be compatible with PGP 5.x.

=item meta_pgp_2_compatible

If true, arguments are generated to try to be compatible with PGP 2.x.

=item meta_interactive

If false, arguments are generated to try to help the using program
use GnuPG in a non-interactive environment, such as CGI scripts.
Default is true.

=item meta_signing_key_id

This scalar reflects the key used to sign messages.
Currently this is synonymous with I<default-key>.

=item meta_signing_key

This GnuPG::Key object reflects the key used to sign messages.

=item meta_recipients_key_ids

This list of scalar key ids are used to generate the
appropriate arguments having these keys as recipients.

=item meta_recipients_keys

This list of keys of the type GnuPG::Key are used to generate the
appropriate arguments having these keys as recipients.
You probably want to have this list be of the inherited class
GnuPG::SubKey, as in most instances, OpenPGP keypairs have
the encyrption key as the subkey of the primary key, which is
used for signing.

=back

=head2 Other Data Members

=over 4

=item extra_args

This is a list of any other arguments used to pass to GnuPG.
Useful to pass an argument not yet covered in this package.

=back

=head1 SEE ALSO

L<GnuPG::Interface>,

=cut
